### PR TITLE
Add breadcrumbs to CSP UI

### PR DESCRIPTION
**Pull request by ari-aviran**
This delivery is the first part of the [benchmarks screen](https://github.com/elastic/security-team/issues/2595), handling the addition of breadcrumbs to our current pages.
Screen Shot 2022-01-04 at 14 18 52

Screen Shot 2022-01-04 at 14 18 48

### CODE REVIEW

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+export const useCspBreadcrumbs = (breadcrumbs: CspNavigationItem[]) => {
+  const {
+    services: {
+      chrome: { setBreadcrumbs },
+      application: { getUrlForApp },
+    },
+  } = useKibana<CoreStart>();
+
+  useEffect(() => {
+    const cspPath = getUrlForApp(PLUGIN_ID);
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:24]: this shouldn't change between runs, maybe define outside of the effect

File: x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+import { i18n } from '@kbn/i18n';
+
+export const CLOUD_POSTURE = i18n.translate('xpack.csp.navigation.cloudPosture', {
+  defaultMessage: 'Cloud Posture',
+});
+
+export const FINDINGS = i18n.translate('xpack.csp.navigation.findings', {
+  defaultMessage: 'Findings',
+});
+
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts:17]: `Findings` page has this text too. i guess this one needs to be in `common` not strictly `navigation`

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    }));
+
+    setBreadcrumbs([
+      {
+        text: CLOUD_POSTURE,
+        href: cspPath,
+      },
+      ...additionalBreadCrumbs,
+    ]);
+  }, [getUrlForApp, setBreadcrumbs, breadcrumbs]);
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:37]: another approach could be listening to `location` changes here, parsing the URL and setting breadcrumbs accordingly. 

i _think_ it'd result in less code, although this is pretty clear, just a bit repetitive.

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * 2.0.
+ */
+
+import * as TEXT from './translations';
+import type { CspNavigationItem } from './types';
+
+export const CSP_FINDINGS_PATH = '/findings';
+export const CSP_DASHBOARD_PATH = '/dashboard';
+
+type NavigableScreens = 'dashboard' | 'findings';
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:14]:
```suggestion
type NavigableScreens = keyof typeof TEXT // can add Lowercase<..> if we want to
```

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+      chrome: { setBreadcrumbs },
+      application: { getUrlForApp },
+    },
+  } = useKibana<CoreStart>();
+
+  useEffect(() => {
+    const cspPath = getUrlForApp(PLUGIN_ID);
+    const additionalBreadCrumbs: ChromeBreadcrumb[] = breadcrumbs.map((breadcrumb) => ({
+      text: breadcrumb.name,
+      path: `${cspPath}${breadcrumb.path}`,
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:27]: we might want to make sure `breadcrumbs.path` starts with `/` or `cspPath` ends with one?

File: x-pack/plugins/cloud_security_posture/public/common/navigation/types.ts [hunk 6]
```diff
@@ -5,9 +5,7 @@
...
 
-import { CSP_FINDINGS_PATH, CSP_DASHBOARD_PATH } from '../common/constants';
-
-export const navigationLinks = [
-  { name: 'Dashboard', path: CSP_DASHBOARD_PATH },
-  { name: 'Findings', path: CSP_FINDINGS_PATH },
-] as const;
+export interface CspNavigationItem {
+  name: string;
+  path: string;
```
Comment (orouz) [inline x-pack/plugins/cloud_security_posture/public/common/navigation/types.ts:10]: not critical, but consider: 

```suggestion
  path: RouteProps["path"];
```

as `path` is going to `Route`

(although that type would be string | string[] so we might need to tweak it more with Extract or casting later on)

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx [hunk 7]
```diff
@@ -11,22 +11,22 @@ import {
...
-      <NavLink to={route.path} activeStyle={activeItemStyle}>
-        {route.name}
-      </NavLink>
-    ),
-  })
-);
+const navItems: NonNullable<KibanaPageTemplateProps['solutionNav']>['items'] = Object.values(
+  allNavigationItems
+).map((route) => ({
+  id: route.name,
```
Comment (JordanSh) [inline x-pack/plugins/cloud_security_posture/public/components/page_template.tsx:22]: `route.name` might get re-evaluated in case of language change. I think using `route.path` is a safer option for something that should be static like an id

**Review by orouz — DISMISSED** [review]
some comments, nothing critical for now

i think later on we might want to change the approach to not require adding breadcrumbs manually but instead automate it for all pages.

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts (hunk 3 above)
Comment (JordanSh) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:37]: I'm also in favor of this method

File: x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_dashboard.tsx [hunk 8]
```diff
@@ -34,12 +36,15 @@ const CompliancePage = () => {
...
-      pageTitle: 'Compliance',
-    }}
-  >
-    <CompliancePage />
-  </CspPageTemplate>
-);
+export const ComplianceDashboard = () => {
+  useCspBreadcrumbs([allNavigationItems.dashboard]);
+  return (
+    <CspPageTemplate
```
Comment (JordanSh) [inline x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_dashboard.tsx:42]:
```suggestion
export const ComplianceDashboard = () => {
  useCspBreadcrumbs([allNavigationItems.dashboard]);
  
  return (
    <CspPageTemplate
```

just added a space, I have a problem with spaces.

Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/pages/compliance_dashboard/compliance_dashboard.tsx:42]: watch out @ari-aviran , it's contagious. i like them spaces now too 😆

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts (hunk 5 above)
Comment (JordanSh) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:27]:
```
`/${string}`
```
🤓

**Review by JordanSh — DISMISSED** [review]

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts (hunk 1 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:24]: I like it this way since its only computed when the effect runs and not when the hook runs (so this is actually computed less times)

File: x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts (hunk 2 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts:17]: This is the same text but has different uses - one is for the navbar/breadcrumbs and one is for the page header which might change

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts (hunk 3 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:37]: Honestly the code in security regarding this is messy and hard to follow. I favor simplicity, when we'll hit trouble we can refactor and abstract

File: x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts (hunk 2 above)
Comment (JordanSh) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts:17]: isn't it the opposite? maybe we want to make sure the breadcrumb title matches the page title?

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx (hunk 7 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.tsx:22]: What do you mean by the second comment? They are statically defined in a constant

File: x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts (hunk 1 above)
Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/use_csp_breadcrumbs.ts:24]: i don't follow how it can be computed less/more, it should be computed once as it doesn't change.  (like put in a ref/state)
although computation here is probably irrelevant, so not critical i guess.

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx (hunk 7 above)
Comment (JordanSh) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.tsx]: I'm not 100% sure how it will work actually with i18, but for example, let's say I'm changing the language mid-session from English to Chinese, we need to make sure the `navItem` rerenders since now the breadcrumbs should be displayed in a different language. Again not completely sure it won't happen anyway, but worth a check.

File: x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts (hunk 2 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/translations.ts:17]: I'll discuss this with Einat and will take care of it in future PRs

**Review by JordanSh — APPROVED** [review]

File: x-pack/plugins/cloud_security_posture/public/components/page_template.tsx (hunk 7 above)
Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.tsx]: You have to refresh for these changes to take effect any way so this is not an issue (since none of the i18n is reactive in any way)

Comment (JordanSh) [reply x-pack/plugins/cloud_security_posture/public/components/page_template.tsx]: yeah i suspected that, cool then no worries about that

File: x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts (hunk 4 above)
Comment (orouz) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:14]: > Using TEXT is not possible since its not typed as constants (they are i18n strings)

not sure i follow. we need the keys not the values. this is what i meant:

in `translations.ts`
```
export const FOO_1 = i18n.translate('key', {
  defaultMessage: 'Some Foo1 Title,
})
export const FOO_2 = i18n.translate('key', {
  defaultMessage: 'Some Foo2 Title,
});
```

in `constants.ts` 

```
import * as TEXT from './constants'

type CspScreens = keyof typeof TEXT // this would be "FOO1 | FOO2"
```

and this way you don't have to manually type `NavigableScreens`, instead you cna derive it from the names of the variables that represent the screen name, as it seems they are the same. (minus being capitalised, which we can fix with `Lowercase`)


> Using keyof on allNavigationItems beats the purpose since I wanted to explicitly type the values there as CspNavigationItem

can you clarify ? not sure i understand

Comment (ari-aviran) [reply x-pack/plugins/cloud_security_posture/public/common/navigation/constants.ts:14]: Oh yea I see about `TEXT` you are correct but this is a wrong coupling IMO, we might have additional navigation related text there in the future.

I don't need `NavigableScreens` as a type, I just wanted the strongest typing for `allNavigationItems`. Some repetition is totally fine IMO (here and in general in coding), especially since the compiler won't let you add a key to `allNavigationItems` without adding an entry in `NavigableScreens`, so you can't "forget"
